File: TandemVec_FCS/analyze_ano.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
AnoCom 姿态数据统计分析（Python 版，numpy + matplotlib 升级）

用法:
    python analyze_ano.py [ano_raw.bin]            # 仅打印文本报告
    python analyze_ano.py ano_raw.bin --plot       # 额外弹出图表
    python analyze_ano.py ano_raw.bin --save fig.png --csv out.csv

解码逻辑与 lib/AnoComProtocol 严格对齐（已与固件逐字节核对）：
    帧结构 : AB | src | dst | func | len(2B小端) | data | SC | AC
    双校验 : SC = 逐字节和 & 0xFF ; AC = "和的和" & 0xFF（uint8 逐步截断）
    0x03   : roll/pitch/yaw = int16 小端 ×100 (度) , fusion @ data[6]
    采样率 : 姿态帧实际间隔 20ms（handleAnoCom@200Hz，4 包分组 → 0x03 为 50Hz）

升级点（相对 JS 版）：
    - numpy 向量化统计 / 校验
    - 漂移率改用最小二乘线性拟合（np.polyfit）+ R²，比首尾差更稳健
    - matplotlib 画 roll/pitch/yaw 时间序列与 yaw 解缠漂移趋势
    - CSV 导出，便于外部工具二次分析
    - 解码 fusionSta：bit7=DETA100在线 / bit6=数据源=DETA100 / 低6位=EKF状态
"""
import sys
import csv
import argparse

try:
    import numpy as np
except ImportError:
    sys.exit("需要 numpy：pip install numpy")

# Windows 控制台默认可能是 GBK，重配 UTF-8 以正常输出中文 / emoji
try:
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
except Exception:
    pass

RANGE_180 = 18000  # ±180° × 100，固件 yaw 约定上限


# ----------------------------- 解码 -----------------------------
def _sum_check(b):
    """SC：逐字节累加 & 0xFF（与固件 uint8 行为一致）。"""
    return int(np.sum(b, dtype=np.int64)) & 0xFF


def _add_check(b):
    """AC：累加和的累加 & 0xFF。cumsum 不逐步截断，但最终 & 0xFF 等价。"""
    return int(np.sum(np.cumsum(b.astype(np.int64)))) & 0xFF


def decode(buf):
    """逐帧解码整段缓冲。CRC 失败逐字节重新同步，而非盲跳整帧。"""
    data = np.frombuffer(buf, dtype=np.uint8)
    n = len(data)
    i = buf.find(b"\xab")
    if i < 0:
        i = n

    total = crc_ok = crc_fail = att_fail = 0
    func_tally = {}
    idx, roll_raw, pitch_raw, yaw_raw, fusion = [], [], [], [], []

    while i + 8 <= n:
        if data[i] != 0xAB:
            i += 1
            continue
        func = int(data[i + 3])
        length = int(data[i + 4]) | (int(data[i + 5]) << 8)
        if length > 256 or i + 8 + length > n:
            i += 1
            continue

        frame = data[i:i + 6 + length]
        ok = (_sum_check(frame) == data[i + 6 + length] and
              _add_check(frame) == data[i + 7 + length])
        if not ok:
            crc_fail += 1
            if func == 0x03:
                att_fail += 1
            i += 1            # 不信任 len，逐字节重新同步
            continue

        # —— CRC 通过，本帧可信 ——
        crc_ok += 1
        total += 1
        func_tally[func] = func_tally.get(func, 0) + 1
        if func == 0x03 and length >= 7:
            p = data[i + 6:i + 6 + length]
            r, pi, y = np.frombuffer(p[:6].tobytes(), dtype="<i2")  # 3×int16 小端
            idx.append(total)
            roll_raw.append(int(r))
            pitch_raw.append(int(pi))
            yaw_raw.append(int(y))
            fusion.append(int(p[6]))
        i += 8 + length       # 仅 CRC 通过才信任 len 跳进整帧

    atts = {
        "idx": np.array(idx, dtype=int),
        "roll_raw": np.array(roll_raw, dtype=int),
        "pitch_raw": np.array(pitch_raw, dtype=int),
        "yaw_raw": np.array(yaw_raw, dtype=int),
        "fusion": np.array(fusion, dtype=int),
    }
    stats = dict(total=total, crc_ok=crc_ok, crc_fail=crc_fail,
                 att_fail=att_fail, func_tally=func_tally)
    return stats, atts


def _unwrap_deg(yaw_deg):
    """对度制 yaw 解缠 ±180° wrap（转弧度借 np.unwrap，兼容所有 numpy 版本）。"""
    return np.degrees(np.unwrap(np.radians(yaw_deg)))


# ----------------------------- 报告 -----------------------------
def report(stats, atts, dt):
    print("\n" + "=" * 30)
    print("  AnoCom 数据统计报告 (Python)")
    print("=" * 30)
    print(f"有效帧: {stats['total']} (CRC通过 {stats['crc_ok']}, CRC失败 {stats['crc_fail']})")
    na = len(atts["idx"])
    extra = f"，另有 {stats['att_fail']} 帧 CRC 失败被丢弃" if stats["att_fail"] else ""
    print(f"姿态帧(0x03): {na}{extra}")
    ft = ", ".join(f"0x{k:02x}={v}" for k, v in sorted(stats["func_tally"].items()))
    print(f"功能码分布: {ft}")

    if na < 2:
        print("\n姿态帧不足，无法分析")
        return

    roll = atts["roll_raw"] / 100.0
    pitch = atts["pitch_raw"] / 100.0
    yaw = atts["yaw_raw"] / 100.0
    yaw_un = _unwrap_deg(yaw)

    # 2. 数据范围健康检查（针对当前固件 ±180° 约定）
    print("\n--- 数据范围健康检查 ---")
    any_out = False
    for name, raw in (("roll ", atts["roll_raw"]),
                      ("pitch", atts["pitch_raw"]),
                      ("yaw  ", atts["yaw_raw"])):
        out = int(np.sum(np.abs(raw) > RANGE_180))
        lo, hi = int(raw.min()), int(raw.max())
        any_out = any_out or out > 0
        flag = f"⚠️  {out} 帧超出 ±180°" if out else "✅"
        print(f"  {name} raw: [{lo}, {hi}]  ({lo/100:.2f}° ~ {hi/100:.2f}°)  {flag}")
    print("  ⚠️  存在超出 ±180° 的样本：固件约定应为 ±180°，请检查 AHRS 输出或 int16 回绕"
          if any_out else
          "  ✅ 全部落在 ±180° 预期范围内（int16 限幅 ±327.67° 远未触及）")

    # 3. 姿态稳定性
    print("\n--- 姿态稳定性 ---")

    def s(arr, name):
        print(f"  {name}: 均值={arr.mean():.3f}° 标准差={arr.std():.4f}° "
              f"范围=[{arr.min():.2f}, {arr.max():.2f}] 波动={arr.max()-arr.min():.3f}°")
    s(roll, "roll ")
    s(pitch, "pitch")
    s(yaw_un, "yaw  ")

    # 4. yaw 漂移率（最小二乘线性拟合 + R²）
    t = np.arange(na) * dt
    slope, intercept = np.polyfit(t, yaw_un, 1)
    fit = slope * t + intercept
    ss_res = float(np.sum((yaw_un - fit) ** 2))
    ss_tot = float(np.sum((yaw_un - yaw_un.mean()) ** 2))
    r2 = 1 - ss_res / ss_tot if ss_tot > 0 else 1.0
    endpoint = float(yaw_un[-1] - yaw_un[0])
    print("\n--- yaw 漂移率 ---")
    print(f"  时长: {t[-1]:.1f}s ({na} 帧 × {dt*1000:.0f}ms 估算)")
    print(f"  线性拟合漂移率: {slope:.4f}°/s  (R²={r2:.3f})")
    print(f"  首尾差参考: {endpoint:.3f}° → {endpoint/t[-1]:.4f}°/s")
    if stats["att_fail"] > 0:
        print(f"  ⚠️  期间 {stats['att_fail']} 个姿态帧 CRC 失败被丢弃，真实时长可能更长，漂移率为偏大估计")
    a = abs(slope)
    print("  ✅ 漂移率极低 (无磁力计静止场景可接受)" if a < 0.05 else
          "  ⚠️  轻微漂移 (无磁力计正常范围)" if a < 0.5 else
          "  ❌ 显著漂移，零偏估计可能未收敛")

    # 5. yaw 突变检测（区分 ±180° wrap 与真实异常）
    print("\n--- yaw 突变检测 ---")
    d = np.diff(yaw)
    big = np.abs(d) > 30
    is_wrap = np.abs(np.abs(d) - 360) < 30
    anomaly = np.where(big & ~is_wrap)[0]
    wraps = int(np.sum(big & is_wrap))
    for j in anomaly:
        print(f"  帧 {j+1}: yaw 从 {yaw[j]:.2f}° 跳到 {yaw[j+1]:.2f}° (Δ={d[j]:.2f}°)")
    if wraps:
        print(f"  ℹ️  {wraps} 次 ±180° 边界 wrap（正常，已在漂移统计中解缠）")
    print("  ✅ 无异常突变" if len(anomaly) == 0 else
          f"  ⚠️  共 {len(anomaly)} 次异常突变（非 wrap），检查 AHRS 输出或丢帧")

    # 6. 融合状态解码（固件 communication.cpp 中 fusionSta 编码）
    print("\n--- 融合状态 (fusionSta 解码) ---")
    fus = atts["fusion"]
    print(f"  DETA100 在线   (bit7): {int(np.sum((fus & 0x80) != 0))}/{na} 帧")
    print(f"  数据源=DETA100 (bit6): {int(np.sum((fus & 0x40) != 0))}/{na} 帧")
    low = fus & 0x3F
    vals, counts = np.unique(low, return_counts=True)
    dist = ", ".join(f"{int(v)}×{int(c)}" for v, c in zip(vals, counts))
    print(f"  EKF 初始化/融合状态 (低6位) 分布: {dist}")

    # 7. 姿态趋势（首尾各 5 帧）
    print("\n--- 姿态趋势 (首5帧) ---")
    for k in range(min(5, na)):
        print(f"  t{atts['idx'][k]:>3}: roll={roll[k]:.2f} pitch={pitch[k]:.2f} "
              f"yaw={yaw[k]:.2f} (raw: {atts['yaw_raw'][k]})")
    print("--- 姿态趋势 (末5帧) ---")
    for k in range(max(0, na - 5), na):
        print(f"  t{atts['idx'][k]:>3}: roll={roll[k]:.2f} pitch={pitch[k]:.2f} "
              f"yaw={yaw[k]:.2f} (raw: {atts['yaw_raw'][k]})")


# ----------------------------- 导出 / 绘图 -----------------------------
def export_csv(atts, dt, path):
    na = len(atts["idx"])
    yaw = atts["yaw_raw"] / 100.0
    yaw_un = _unwrap_deg(yaw) if na >= 2 else yaw
    with open(path, "w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow(["idx", "t_s", "roll_deg", "pitch_deg", "yaw_deg",
                    "yaw_unwrap_deg", "roll_raw", "pitch_raw", "yaw_raw", "fusion"])
        for k in range(na):
            w.writerow([int(atts["idx"][k]), round(k * dt, 4),
                        atts["roll_raw"][k] / 100, atts["pitch_raw"][k] / 100,
                        atts["yaw_raw"][k] / 100, round(float(yaw_un[k]), 4),
                        int(atts["roll_raw"][k]), int(atts["pitch_raw"][k]),
                        int(atts["yaw_raw"][k]), int(atts["fusion"][k])])
    print(f"\n已导出 CSV: {path}  ({na} 行)")


def plot(atts, dt, save=None, show=False):
    na = len(atts["idx"])
    if na < 2:
        print("  姿态帧不足，跳过绘图")
        return
    try:
        import matplotlib
        if save and not show:
            matplotlib.use("Agg")     # 无 GUI 也能保存
        import matplotlib.pyplot as plt
    except ImportError:
        print("  未安装 matplotlib，跳过绘图：pip install matplotlib")
        return

    roll = atts["roll_raw"] / 100.0
    pitch = atts["pitch_raw"] / 100.0
    yaw = atts["yaw_raw"] / 100.0
    yaw_un = _unwrap_deg(yaw)
    t = np.arange(na) * dt
    slope, intercept = np.polyfit(t, yaw_un, 1)

    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(11, 7), sharex=True)
    ax1.plot(t, roll, lw=1, label="roll")
    ax1.plot(t, pitch, lw=1, label="pitch")
    ax1.plot(t, yaw, lw=1, label="yaw")
    ax1.set_ylabel("angle (deg)")
    ax1.set_title("AnoCom attitude time series")
    ax1.legend(loc="upper right")
    ax1.grid(alpha=.3)

    ax2.plot(t, yaw_un, lw=1, color="tab:red", label="yaw (unwrapped)")
    ax2.plot(t, slope * t + intercept, "--", color="k", lw=1,
             label=f"fit {slope:.4f} deg/s")
    ax2.set_xlabel("time (s)")
    ax2.set_ylabel("yaw unwrapped (deg)")
    ax2.set_title("yaw drift trend")
    ax2.legend(loc="upper left")
    ax2.grid(alpha=.3)

    fig.tight_layout()
    if save:
        fig.savefig(save, dpi=120)
        print(f"  图表已保存: {save}")
    if show:
        plt.show()


# ----------------------------- 入口 -----------------------------
def main():
    ap = argparse.ArgumentParser(description="AnoCom 姿态数据统计分析 (Python 版)")
    ap.add_argument("file", nargs="?", default="ano_raw.bin", help="二进制 dump 文件")
    ap.add_argument("--dt", type=float, default=0.02,
                    help="姿态帧间隔(秒)，默认 0.02 (50Hz)")
    ap.add_argument("--plot", "-p", action="store_true", help="弹出图表窗口")
    ap.add_argument("--save", metavar="PNG", help="保存图表到 PNG 文件")
    ap.add_argument("--csv", metavar="CSV", help="导出姿态数据到 CSV")
    args = ap.parse_args()

    try:
        with open(args.file, "rb") as f:
            buf = f.read()
    except OSError as e:
        sys.exit(f'无法读取文件 "{args.file}": {e}')
    print(f"读取 {len(buf)} 字节 ({args.file})")

    stats, atts = decode(buf)
    report(stats, atts, args.dt)

    if args.csv:
        export_csv(atts, args.dt, args.csv)
    if args.plot or args.save:
        plot(atts, args.dt, save=args.save, show=args.plot)


if __name__ == "__main__":
    main()

File: TandemVec_FCS/analyze_ano.js
// AnoCom 姿态数据统计分析脚本
// 用法: node analyze_ano.js [ano_raw.bin]
// 解析固件 AnoComProtocol 发送的二进制帧流 (帧头 0xAB)，统计姿态稳定性 / 漂移 / 异常。
// 协议与 lib/AnoComProtocol 严格对齐：2 字节小端长度、双校验(SC+AC)、0x03 姿态帧 int16×100。
const fs = require('fs');

// ---- 文件读取（带友好报错） ----
const path = process.argv[2] || 'ano_raw.bin';
let buf;
try {
  buf = fs.readFileSync(path);
} catch (e) {
  console.error(`无法读取文件 "${path}": ${e.message}`);
  console.error(`用法: node analyze_ano.js [ano_raw.bin]`);
  process.exit(1);
}
console.log(`读取 ${buf.length} 字节 (${path})`);

// ---- 基础解码函数 ----
// 用乘法而非 <<，避免 4 字节最高位导致的符号问题；les 统一处理有符号
const le  = (d,o,n)=>{let v=0;for(let k=0;k<n;k++)v+=d[o+k]*(2**(8*k));return v;};
const les = (d,o,n)=>{let v=le(d,o,n);const ms=2**(8*n-1);if(v>=ms)v-=2**(8*n);return v;};
// 逐步 &0xFF，与固件 uint8_t 累加行为完全一致
function sumCheck(d,l){let s=0;for(let i=0;i<l;i++)s=(s+d[i])&0xFF;return s;}
function addCheck(d,l){let s=0,a=0;for(let i=0;i<l;i++){s=(s+d[i])&0xFF;a=(a+s)&0xFF;}return a;}

// ---- 数组统计辅助（循环求值，避免 Math.min(...arr) 对大数组栈溢出） ----
function minOf(arr){let m=Infinity;for(const v of arr)if(v<m)m=v;return m;}
function maxOf(arr){let m=-Infinity;for(const v of arr)if(v>m)m=v;return m;}

// ---- 逐帧解码（无帧数上限，分析整个文件） ----
let i = Math.max(0, buf.indexOf(0xAB));
const atts = [];
let totalFrames=0, crcOk=0, crcFail=0, attFail=0;
const funcTally = {};

while (i + 8 <= buf.length) {
  if (buf[i] !== 0xAB) { i++; continue; }
  const func = buf[i+3], len = buf[i+4] | (buf[i+5]<<8);
  // 长度非法或越界：当前位置不是合法帧头，逐字节重新同步
  if (len > 256 || i+8+len > buf.length) { i++; continue; }

  const cr = buf.slice(i, i+6+len);
  const isCrcOk = sumCheck(cr,len+6)===buf[i+6+len] && addCheck(cr,len+6)===buf[i+7+len];

  if (!isCrcOk) {
    // CRC 失败：可能是错位/噪声，不信任 len，逐字节重新同步（而非跳过整帧）
    crcFail++;
    if (func === 0x03) attFail++;
    i++;
    continue;
  }

  // —— CRC 通过，本帧可信 ——
  crcOk++;
  totalFrames++;
  funcTally[func] = (funcTally[func]||0)+1;

  if (func===0x03 && len>=7) {
    const p = buf.slice(i+6, i+6+len);
    const roll_raw  = les(p,0,2);  // int16, x100 deg
    const pitch_raw = les(p,2,2);
    const yaw_raw   = les(p,4,2);
    atts.push({
      idx: totalFrames,
      roll: roll_raw/100, pitch: pitch_raw/100, yaw: yaw_raw/100,
      roll_raw, pitch_raw, yaw_raw,
      fusion: p[6]
    });
  }
  i = i + 8 + len;  // 仅 CRC 通过才信任 len 跳进整帧
}

// ---- 1. 总体统计 ----
console.log(`\n==============================`);
console.log(`  AnoCom 数据统计报告`);
console.log(`==============================`);
console.log(`有效帧: ${totalFrames} (CRC通过 ${crcOk}, CRC失败 ${crcFail})`);
console.log(`姿态帧(0x03): ${atts.length}${attFail?`，另有 ${attFail} 帧 CRC 失败被丢弃`:''}`);
console.log(`功能码分布:`, Object.entries(funcTally).map(([k,v])=>`0x${(+k).toString(16).padStart(2,'0')}=${v}`).join(', '));

if(atts.length < 2) { console.log('\n姿态帧不足，无法分析'); process.exit(0); }

// ---- 2. 数据范围健康检查 ----
// 固件 sendAttitudeEuler 用 (int16_t)(角度×100) 编码，理论限幅 ±327.67°。
// 但 communication.cpp 已将 yaw wrap 到 ±180°，roll/pitch 物理上也在 ±180° 内，
// 因此正常情况下 raw 恒在 ±18000，绝不会触及 int16 限幅。
// 这里检测真正可能发生的异常：raw 超出 ±180° 预期范围（AHRS 异常 / 约定变更 / int16 回绕征兆）。
console.log(`\n--- 数据范围健康检查 ---`);
const RANGE_180 = 18000; // ±180° × 100
const axes = [
  ['roll ', atts.map(a=>a.roll_raw)],
  ['pitch', atts.map(a=>a.pitch_raw)],
  ['yaw  ', atts.map(a=>a.yaw_raw)],
];
let anyOut = false;
for(const [name, raw] of axes){
  const lo = minOf(raw), hi = maxOf(raw);
  let out = 0; for(const v of raw) if(Math.abs(v) > RANGE_180) out++;
  if(out>0) anyOut = true;
  const flag = out>0 ? `⚠️  ${out} 帧超出 ±180°` : '✅';
  console.log(`  ${name} raw: [${lo}, ${hi}]  (${(lo/100).toFixed(2)}° ~ ${(hi/100).toFixed(2)}°)  ${flag}`);
}
if(anyOut) console.log(`  ⚠️  存在超出 ±180° 的样本：固件约定应为 ±180°，请检查 AHRS 输出或 int16 回绕`);
else       console.log(`  ✅ 全部落在 ±180° 预期范围内（int16 限幅 ±327.67° 远未触及）`);

// ---- 3. 姿态稳定性统计 ----
console.log(`\n--- 姿态稳定性 ---`);
const stats = (arr, name) => {
  const min = minOf(arr), max = maxOf(arr);
  const range = max - min;
  let sum=0; for(const v of arr) sum+=v;
  const mean = sum/arr.length;
  let varSum=0; for(const v of arr) varSum+=(v-mean)**2;
  const stddev = Math.sqrt(varSum/arr.length);
  console.log(`  ${name}: 均值=${mean.toFixed(3)}° 标准差=${stddev.toFixed(4)}° 范围=[${min.toFixed(2)}, ${max.toFixed(2)}] 波动=${range.toFixed(3)}°`);
};
stats(atts.map(a=>a.roll), 'roll ');
stats(atts.map(a=>a.pitch), 'pitch');
// yaw 需要处理 wrap-around（±180° 边界）
const yawUnwrapped = [atts[0].yaw];
for(let j=1; j<atts.length; j++){
  let diff = atts[j].yaw - atts[j-1].yaw;
  if(diff > 180) diff -= 360;
  if(diff < -180) diff += 360;
  yawUnwrapped.push(yawUnwrapped[j-1] + diff);
}
stats(yawUnwrapped, 'yaw  ');

// ---- 4. yaw 漂移率 ----
// 姿态帧实际间隔：handleAnoCom @200Hz，4 包分组，0x03 每 4 次发一次 → 50Hz / 20ms
const DT = 0.02;
const dt_total = (atts.length - 1) * DT;
const yaw_drift_total = yawUnwrapped[yawUnwrapped.length-1] - yawUnwrapped[0];
const yaw_drift_rate = yaw_drift_total / dt_total;
console.log(`\n--- yaw 漂移率 ---`);
console.log(`  总变化: ${yaw_drift_total.toFixed(3)}° over ${dt_total.toFixed(1)}s (按 ${atts.length} 帧 × 20ms 估算)`);
console.log(`  漂移率: ${yaw_drift_rate.toFixed(4)}°/s`);
if(attFail>0)
  console.log(`  ⚠️  期间有 ${attFail} 个姿态帧 CRC 失败被丢弃，真实时长可能更长，漂移率为偏大估计`);
if(Math.abs(yaw_drift_rate) < 0.05) console.log(`  ✅ 漂移率极低 (无磁力计静止场景可接受)`);
else if(Math.abs(yaw_drift_rate) < 0.5) console.log(`  ⚠️  轻微漂移 (无磁力计正常范围)`);
else console.log(`  ❌ 显著漂移，零偏估计可能未收敛`);

// ---- 5. yaw 突变检测 ----
// 相邻原始 yaw 跳变 >30°：跨 ±180° 边界的 wrap (|Δ|≈360) 是正常的，其余才是异常
console.log(`\n--- yaw 突变检测 ---`);
let jumps = 0, wraps = 0;
for(let j=1; j<atts.length; j++){
  const diff = atts[j].yaw - atts[j-1].yaw;
  if(Math.abs(diff) > 30) {
    if(Math.abs(Math.abs(diff) - 360) < 30) { wraps++; continue; } // |Δ|≈360 → ±180 边界 wrap
    jumps++;
    console.log(`  帧 ${j}: yaw 从 ${atts[j-1].yaw.toFixed(2)}° 跳到 ${atts[j].yaw.toFixed(2)}° (Δ=${diff.toFixed(2)}°)`);
  }
}
if(wraps>0) console.log(`  ℹ️  ${wraps} 次 ±180° 边界 wrap（正常，已在漂移统计中解缠）`);
if(jumps === 0) console.log(`  ✅ 无异常突变`);
else console.log(`  ⚠️  共 ${jumps} 次异常突变（非 wrap），检查 AHRS 输出或丢帧`);

// ---- 6. 数据趋势（首5帧 + 末5帧） ----
console.log(`\n--- 姿态趋势 (首5帧) ---`);
atts.slice(0,5).forEach(a =>
  console.log(`  t${String(a.idx).padStart(3)}: roll=${a.roll.toFixed(2)} pitch=${a.pitch.toFixed(2)} yaw=${a.yaw.toFixed(2)} (raw: ${a.yaw_raw})`)
);
console.log(`--- 姿态趋势 (末5帧) ---`);
atts.slice(-5).forEach(a =>
  console.log(`  t${String(a.idx).padStart(3)}: roll=${a.roll.toFixed(2)} pitch=${a.pitch.toFixed(2)} yaw=${a.yaw.toFixed(2)} (raw: ${a.yaw_raw})`)
);
